lib/matplotlib/dates.py

```diff
@@ -230,7 +230,7 @@ def _get_tzinfo(tz=None):
         return tzinfo
     if isinstance(tz, datetime.tzinfo):
         return tz
-    raise TypeError("tz must be string or tzinfo subclass.")
+    raise TypeError(f"tz must be string or tzinfo subclass, not {tz!r}.")
```

Comment:
I think this is good, but perhaps more fundamentally is that dates.py does not define allowable units...

Reply:
Yes, there seems to be a bit of "abuse" of things. This line doesn't really look like one would hope for:
https://github.com/matplotlib/matplotlib/blob/75a35e0fedaadd49f7b3a2015a1ced6f426a1b69/lib/matplotlib/dates.py#L1798

To me it seems like one is patching dates to, sort of, work with units. But then I have very limited insights into how the units do work.

There can for sure be better fixes for this, but this one the one I could add to at least improve the situation slightly...

My interpretation of the problem is that there is no checking for validity when setting the units. Then, when setting timezone through `axis_date`, whatever is set as `self.units` of the `Axis` is used and an `AutoDateLocator` is instantiated with that as `tz`...